sympy/tensor/tests/test_tensor.py

```diff
@@ -2040,3 +2040,22 @@ def test_TensorType():
 def test_dummy_fmt():
     with warns_deprecated_sympy():
         TensorIndexType('Lorentz', dummy_fmt='L')
+
+def test_postprocessor():
+    """
+    Test if substituting a Tensor into a Mul or Add automatically converts it
+    to TensMul or TensAdd respectively. See github issue #25051
+    """
+    R3 = TensorIndexType('R3', dim=3)
+    i = tensor_indices("i", R3)
+    K = TensorHead("K", [R3])
+    x,y,z = symbols("x y z")
+
+    assert isinstance((x*2).xreplace({x: K(i)}), TensMul)
+    assert isinstance((x+2).xreplace({x: K(i)*K(-i)}), TensAdd)
+
+    assert isinstance((x*2).subs({x: K(i)}), TensMul)
+    assert isinstance((x+2).subs({x: K(i)*K(-i)}), TensAdd)
+
+    assert isinstance((x*2).replace(x, K(i)), TensMul)
+    assert isinstance((x+2).replace(x, K(i)*K(-i)), TensAdd)
```

Comment:
have you tried a higher level of nestedness? For example, `(2 + x*y).replace(x, ... )` ?

Reply:
It seems to work fine with the changes in this PR:
```
In [13]: type((x*2 + 4).replace(x, K(p)*K(-p)))
Out[13]: sympy.tensor.tensor.TensAdd

In [14]: type((x*y + 4).replace(x, K(p)*K(-p)))
Out[14]: sympy.tensor.tensor.TensAdd

In [15]: type((z*(x*y + 4)).replace(x, K(p)*K(-p)))
Out[15]: sympy.tensor.tensor.TensMul
```

Do you want me to add these as tests?